airflow/models/base_async_operator.py

```diff
@@ -0,0 +1,96 @@
+# specific language governing permissions and limitations
+# under the License.
+"""
+Base Asynchronous Operator for kicking off a long running
+operations and polling for completion with reschedule mode.
+"""
+from functools import wraps
+from airflow.sensors.base_sensor_operator import BaseSensorOperator
+from airflow.exceptions import AirflowException
+from airflow.models.xcom import XCOM_EXTERNAL_RESOURCE_ID_KEY
+
+class BaseAsyncOperator(BaseSensorOperator, SkipMixin):
+    """
+    AsyncOperators are derived from this class and inherit these attributes.
+
+    AsyncOperators must define a `submit_request` to fire a request for a
+    long running operation with a method and then executes a `poke` method
+    executing at a time interval and succeed when a criteria is met and fail
+    if and when they time out. They are effctively an opinionated way use
+    combine an Operator and a Sensor in order to kick off a long running
+    process without blocking a worker slot while waiting for the long running
+    process to complete by leveraging reschedule mode.
+
+    :param soft_fail: Set to true to mark the task as SKIPPED on failure
+    :type soft_fail: bool
+    :param poke_interval: Time in seconds that the job should wait in
+        between each tries
+    :type poke_interval: int
+    :param timeout: Time, in seconds before the task times out and fails.
+    :type timeout: int
+    :type mode: str
+    """
+    ui_color = '#9933ff'  # type: str
+    valid_modes = ['poke', 'reschedule']  # type: Iterable[str]
+
+    @apply_defaults
+    def __init__(self,
+                 *args,
+                 **kwargs) -> None:
+        super().__init__(mode='reschedule', *args, **kwargs)
+
+    def submit_request(self, context) -> string:
+        """
+        This method should kick off a long running operation.
+        This method should return the ID for the long running operation used
+        for polling
+        Context is the same dictionary used as when rendering jinja templates.
+
+        Refer to get_template_context for more context.
+
+        :returns: a resource_id for the long running operation.
+        :rtype: str
+        """
+        raise AirflowException('Async Operators must define a `submit_request` method.')
+
+    def process_result(self, context):
+        """
+        This method can optionally be overriden to process the result of a long running operation.
+        Context is the same dictionary used as when rendering jinja templates.
+
+        Refer to get_template_context for more context.
+        """
+        self.log.info('Got result of {}. Done.'.format(
+                      self.get_external_resource_id(context))
+
+    def pre_execute(self, context) -> None:
+        """
+        Check if we have the XCOM_EXTERNAL_RESOURCE_ID_KEY
+        for this task and call submit_request if it is missing.
+        """
+        if not self.get_external_resource_id(context):
```

Comment:
I think that we should not rely on the content of this field. I can imagine that there may be operators that do not operate on any identifier.

Reply:
If an operator doesn't operate on an identifier what is there to poke? Maybe we can have BaseAsyncOperator set a default `'PLACEHOLDER_ID'` ?

Reply:
There can only be one object in the universe.  In this case, there is a constant question about a unique object. After thinking for a long time, however, I have concerns about performance. Now we have one query to the database, and if we tried to save the state then there would always be two queries.

Reply:
Apologies, I'm really not following you here.

By "one object in the universe" do you mean there can only be one external resource?

Why is the object uniqueness in question?

Can you explain the performance concern? performance of what? the task? the submit? I'm envisioning this async operator pattern only being used for long running operations where you can tolerate a poke interval of >1 min.